### PR TITLE
Add MacOS C++ CI, add Linux C++20 CI

### LINKED ISSUE #4116: Add CI step for building C++ on Mac

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes #4116

New way of doing test matrices: Load different json files to configure an entire job.
-> Much nicer config interface (a separate clean file <3) and less & less arcane job setup \o/

Adds tests for Mac AND Linux C++20!

Runs full matrix on main and whenever cpp/hpp/cmake changes are registered.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/4120) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4120)
- [Docs preview](https://rerun.io/preview/0f7f7ebe692424cf6a9e4f269f3b0790c74f9acd/docs) 
- [Examples preview](https://rerun.io/preview/0f7f7ebe692424cf6a9e4f269f3b0790c74f9acd/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: .github/workflows/on_pull_request.yml [hunk 1]
```diff
@@ -22,6 +22,31 @@ jobs:
...
+    outputs:
+      cpp_changes: ${{ steps.filter.outputs.rerun_cpp }}
+    steps:
+    - uses: actions/checkout@v2
+    - uses: dorny/paths-filter@v2
+      id: filter
+      with:
+        filters: |
+          rerun_cpp:
+            - 'rerun_cpp/**'
```
Comment (emilk) [inline .github/workflows/on_pull_request.yml:37]: This misses a lot of stuff. Can we do `**/*.hpp`, `**/.cpp`, and `**/CMakeLists.txt` instead?

File: .github/workflows/on_pull_request.yml [hunk 2]
```diff
@@ -22,6 +22,31 @@ jobs:
...
+
+  # Run cpp tests for all platforms
+  cpp-tests:
+    name: "C++ tests"
+    # Wait from the cpp-paths-filter to be completed before starting next-job
+    needs: cpp-paths-filter
```
Comment (emilk) [inline .github/workflows/on_pull_request.yml:43]:
```suggestion
    # Wait for the cpp-paths-filter to be completed before starting next-job
    needs: cpp-paths-filter
```

File: .github/workflows/on_push_main.yml [hunk 3]
```diff
@@ -163,7 +170,7 @@ jobs:
     secrets: inherit
 
   build-rerun-cli-and-upload-windows:
-    needs: [checks]
+    needs: [checks, checks_cpp]
```
Comment (emilk) [inline .github/workflows/on_push_main.yml:173]: This creates pipelining, adding end-to-end build time.

Why? Are these jobs using some output from `checks_cpp`?

File: .github/workflows/reusable_build_and_upload_rerun_cli.yml [hunk 4]
```diff
@@ -59,7 +59,7 @@ permissions:
 jobs:
   set-config:
     name: Set Config (${{ inputs.PLATFORM }})
-    runs-on: ubuntu-latest-16-cores
+    runs-on: ubuntu-latest
```
Comment (emilk) [inline .github/workflows/reusable_build_and_upload_rerun_cli.yml:62]: Was it intentional to nerf this runner?

File: .github/workflows/reusable_checks_cpp.yml [hunk 5]
```diff
@@ -0,0 +1,85 @@
+name: "C++ Tests on all platforms & compilers"
+
+on:
+  workflow_call:
+    inputs:
+      CONCURRENCY:
+        required: true
+        type: string
+      MINIMAL:
```
Comment (emilk) [inline .github/workflows/reusable_checks_cpp.yml:9]: nit: Maybe call this `FAST` to better indicate its intention, or flip it and call it `FULL` 🤷

File: .github/workflows/reusable_checks_cpp.yml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.ref || '' }}
+      - name: Load minimal C++ test matrix
+        if: ${{ github.event.inputs.MINIMAL }}
+        run: echo "matrix=$(jq -c . < ./.github/cpp_matrix_full.json)" >> $GITHUB_OUTPUT
+      - name: Load full C++ test matrix
+        if: ${{ !github.event.inputs.MINIMAL }}
+        run: echo "matrix=$(jq -c . < ./.github/cpp_matrix_minimal.json)" >> $GITHUB_OUTPUT
```
Comment (emilk) [inline .github/workflows/reusable_checks_cpp.yml]: It is good practice to use proper antonyms for opposites, and I'm not sure "full" and "minimal" qualifies. Maybe:

* partial / full
* minimal / maximal
* some / all
* fast / slow

**Review by emilk — APPROVED** [review]
Nice to get this refactored into a proper matrix!

File: .github/workflows/reusable_checks_cpp.yml [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.ref || '' }}
+      - name: Load minimal C++ test matrix
+        if: ${{ github.event.inputs.MINIMAL }}
+        run: echo "matrix=$(jq -c . < ./.github/workflows/cpp_matrix_full.json)" >> $GITHUB_OUTPUT
+      - name: Load full C++ test matrix
+        if: ${{ !github.event.inputs.MINIMAL }}
+        run: echo "matrix=$(jq -c . < ./.github/workflows/cpp_matrix_minimal.json)" >> $GITHUB_OUTPUT
```
Comment (emilk) [inline .github/workflows/reusable_checks_cpp.yml:51]: It is good practice to use proper antonyms for opposites, and I'm not sure "full" and "minimal" qualifies. Maybe:

* partial / full
* minimal / maximal
* some / all
* fast / slow

**Comment by Wumpf** [comment]
@emilk this is still under development, nothing is working yet ;)

File: .github/workflows/reusable_build_and_upload_rerun_cli.yml (hunk 4 above)
Comment (Wumpf) [reply .github/workflows/reusable_build_and_upload_rerun_cli.yml:62]: yeah it just sets config, doesn't make sense for it to have more cores I think?

File: .github/workflows/on_push_main.yml (hunk 3 above)
Comment (Wumpf) [reply .github/workflows/on_push_main.yml:173]: no, but they also don't use anything from checks do they? Thought this is so we don't do unnecssary work if stuff fails

File: .github/workflows/on_pull_request.yml (hunk 1 above)
Comment (Wumpf) [reply .github/workflows/on_pull_request.yml:37]: I have no idea if this works anyways at this point

**Review by jprochazk — APPROVED** [review]

**Pull request merged**